mp_sqrtmod_prime.c

```diff
@@ -13,28 +13,30 @@ mp_err mp_sqrtmod_prime(const mp_int *n, const mp_int *prime, mp_int *ret)
 {
    mp_err err;
    int legendre;
-   mp_int t1, C, Q, S, Z, M, T, R, two;
-   mp_digit i;
+   /* The type is "int" because of the types in the mp_int struct.
+      Don't forget to change them here when you change them there! */
+   int S, M, i;
+   mp_int t1, C, Q, Z, T, R, two;
 
    /* first handle the simple cases */
    if (mp_cmp_d(n, 0uL) == MP_EQ) {
       mp_zero(ret);
       return MP_OKAY;
    }
-   if (mp_cmp_d(prime, 2uL) == MP_EQ)                            return MP_VAL; /* prime must be odd */
-   if ((err = mp_kronecker(n, prime, &legendre)) != MP_OKAY)        return err;
+   if (mp_iseven(prime) || (mp_cmp_d(prime, 3uL) == MP_LT))      return MP_VAL; /* "prime" must be odd and > 2 */
+   if ((err = mp_kronecker(n, prime, &legendre)) != MP_OKAY)     return err;
    if (legendre == -1)                                           return MP_VAL; /* quadratic non-residue mod prime */
```

Comment:
Presumably a user might check that gcd(n, prime) = 1 before running this function, but might be better to change this line to
```C
if (legendre != 1) return MP_VAL;
```